packages/core/lib/commands/migrate/runMigrations.js

```diff
@@ -1,6 +1,110 @@
+const { createReadStream } = require("fs");
+const { readdir } = require("fs/promises");
+const path = require("path");
+const JSONStream = require("JSONStream");
 const Migrate = require("@truffle/migrate").default;
+const TruffleError = require("@truffle/error");
+const os = require("os");
+const debug = require("debug")("migrate:run");
+
+// Search for Push1 (60) to Push(32) 7F + console.log
+//                      60 ---- 7F  C O N S O L E . L O G
+const consoleLogRex = /[67][0-9a-f]636F6e736F6c652e6c6f67/i;
+
+async function usesConsoleLog(artifactJson) {
+  const debugLog = debug.extend("test");
+  debugLog("Artifact: %o", artifactJson);
+
+  //create a parser to get the value of jsonpath .deployedBytecode
+  const parser = JSONStream.parse(["deployedBytecode"]);
+  const stream = createReadStream(artifactJson).pipe(parser);
+
+  return new Promise((resolve, reject) => {
+    stream.on("data", data => {
+      //JSONParse will emit the entire string/value
+      //so initiate stream cleanup here
+      stream.destroy();
+      const usesConsoleLog = consoleLogRex.test(data);
+      debugLog("usesConsoleLog:", usesConsoleLog);
+      resolve(usesConsoleLog);
+    });
+
+    stream.on("error", err => {
+      stream.destroy();
+      debugLog("onError: %o", err);
+      reject(err);
+    });
+  });
+}
+
+async function findArtifactsThatUseConsoleLog(buildDir) {
+  const debugLog = debug.extend("dirty-files");
+  const files = await readdir(buildDir);
+
+  const artifacts = [];
+  await Promise.allSettled(
```

Comment:
You can't really use an `async` function with `forEach`. I mean, you can but it is probably not doing what you think it is doing.

I'm pretty sure `forEach` doesn't "respect" `async` functions in that it will execute each one and the code flow will continue on without waiting for the `async` stuff to resolve. `for` loops, however, do "respect" `await`s :)

Reply:
@eggplantzzz good point, I was just quickly writing up an example and you are right that you wouldn't want to use `foreach` if needing to wait for each line to resolve before continuing. I guess I'd still prefer filter and then a for loop because it is more readable but I get @cds-amal's point re: looping through twice vs. once and this is an efficient way to use `Promise.allSettled()` + `map`.